File: python/periph/chips/tof/_vl53_base.py
"""Shared base for ST's VL53 FlightSense Time-of-Flight ranging family.

Internal module — not a public driver. VL53L0X and VL53L1X have different
register maps (8-bit vs 16-bit register index), so this base holds no
register addresses and no ranging logic, only the plumbing both chips share:
big-endian register access with a 1- or 2-byte index, the bounded poll
helper, the XSHUT boot wait, the lock that serializes multi-register
sequences against the interrupt poller, interrupt delivery (int_pin edge or
Linux polling thread), the volatile re-addressing helper, and the family's
shared constants. See specs/tof/_vl53_base.md.
"""

try:
    import threading as _threading
    _LINUX = True
except ImportError:
    _LINUX = False

import time


I2C_ADDRESS = 0x29

TIMEOUT_MS = 500
BOOT_US = 1200

# Logical interrupt sources shared by the family (mutually exclusive).
SOURCE_LEVEL_LOW        = 1
SOURCE_LEVEL_HIGH       = 2
SOURCE_OUT_OF_WINDOW    = 3
SOURCE_NEW_SAMPLE_READY = 4
SOURCE_IN_WINDOW        = 5  # VL53L1X only


class _NoLock:
    def __enter__(self):
        return self

    def __exit__(self, *exc):
        return False


def _ticks_ms():
    if hasattr(time, 'ticks_ms'):
        return time.ticks_ms()
    if hasattr(time, 'monotonic'):
        return int(time.monotonic() * 1000)
    return int(time.time() * 1000)


def _elapsed_ms(start):
    if hasattr(time, 'ticks_diff'):
        return time.ticks_diff(time.ticks_ms(), start)
    return _ticks_ms() - start


class _VL53Base:
    """Protected helpers shared by the VL53 chip drivers.

    Never instantiated directly. Subclasses implement
    _poll_interrupt_status() for the interrupt delivery helpers.

    Args:
        connection: Configured I2C connection. Its optional en_pin drives
            XSHUT (high = enabled), its optional int_pin receives GPIO1.
        index_bytes: Register index width on the wire, 1 or 2.
        chip_name: Chip name used in error messages.
    """

    def __init__(self, connection, index_bytes, chip_name):
        self._connection = connection
        self._index_bytes = index_bytes
        self._chip_name = chip_name
        # Serializes multi-register sequences against the interrupt polling
        # thread (a status read/clear must not interleave with them).
        self._lock = _threading.RLock() if _LINUX else _NoLock()
        self._callback = None
        self._int_pin_used = None
        self._poll_stop = False
        self._poll_thread = None

    # --- Register access --------------------------------------------------

    def _index(self, reg):
        if self._index_bytes == 2:
            return bytes([(reg >> 8) & 0xFF, reg & 0xFF])
        return bytes([reg & 0xFF])

    def _wr_block(self, reg, data):
        self._connection.write(self._index(reg) + bytes(data))

    def _rd_block(self, reg, n):
        return self._connection.write_read(self._index(reg), n)

    def _wr8(self, reg, value):
        self._wr_block(reg, (value & 0xFF,))

    def _rd8(self, reg):
        return self._rd_block(reg, 1)[0]

    def _wr16(self, reg, value):
        self._wr_block(reg, ((value >> 8) & 0xFF, value & 0xFF))

    def _rd16(self, reg):
        data = self._rd_block(reg, 2)
        return (data[0] << 8) | data[1]

    def _wr32(self, reg, value):
        self._wr_block(reg, ((value >> 24) & 0xFF, (value >> 16) & 0xFF,
                             (value >> 8) & 0xFF, value & 0xFF))

    def _rd32(self, reg):
        data = self._rd_block(reg, 4)
        return (data[0] << 24) | (data[1] << 16) | (data[2] << 8) | data[3]

    # --- Polling and boot -------------------------------------------------

    def _wait_until(self, predicate, what):
        start = _ticks_ms()
        while True:
            if predicate():
                return
            if _elapsed_ms(start) > TIMEOUT_MS:
                raise OSError('{} timeout waiting for {}'.format(self._chip_name, what))

    def _boot_wait(self):
        if getattr(self._connection, 'en_pin', None):
            self._connection.enable()
        time.sleep(BOOT_US / 1000000)

    # --- Re-addressing ----------------------------------------------------

    def _set_address_reg(self, reg, address):
        if address < 0x08 or address > 0x77:
            raise ValueError('address must be 0x08 to 0x77')
        self._wr8(reg, address & 0x7F)

    # --- Interrupt delivery -----------------------------------------------

    def _poll_interrupt_status(self):
        """Read and clear a pending interrupt; return its SOURCE_* value or 0."""
        raise NotImplementedError

    def _subscribe(self, callback, int_pin=None):
        self._callback = callback
        pin = int_pin if int_pin is not None else getattr(self._connection, 'int_pin', None)
        self._int_pin_used = pin
        if pin is not None:
            from periph.connection.input_pin import InputPin
            pin.on_edge(self._int_handler, InputPin.FALLING)
        elif _LINUX:
            self._poll_stop = False
            self._poll_thread = _threading.Thread(target=self._poll_loop, daemon=True)
            self._poll_thread.start()

    def _unsubscribe(self):
        if self._int_pin_used is not None:
            self._int_pin_used.off_edge(self._int_handler)
            self._int_pin_used = None
        elif _LINUX:
            self._poll_stop = True
        self._callback = None

    def _int_handler(self):
        status = self._poll_interrupt_status()
        if status and self._callback:
            self._callback(status)

    def _poll_loop(self):
        while not self._poll_stop:
            status = self._poll_interrupt_status()
            if status and self._callback:
                self._callback(status)
            time.sleep(0.005)

File: go/periph/chips/tof/vl53_base.go
package tof

import (
	"sync"
	"time"

	"github.com/tuhde/Periph/go/periph/connection"
)

// Shared plumbing for ST's VL53 FlightSense Time-of-Flight ranging family.
// VL53L0X and VL53L1X have different register maps (8-bit vs 16-bit register
// index), so vl53Base holds no register addresses and no ranging logic, only
// what both chips share: big-endian register access with a 1- or 2-byte
// index, the bounded poll helper, the XSHUT boot wait, the mutex that
// serializes multi-register sequences against the interrupt goroutine,
// interrupt delivery (IntPin edge or 5 ms polling pin), and the volatile
// re-addressing helper. It is embedded by value in each chip's Minimal
// driver; see specs/tof/_vl53_base.md.

const (
	vl53I2CAddress uint8 = 0x29
	vl53Timeout          = 500 * time.Millisecond
	vl53BootTime         = 1200 * time.Microsecond

	// Logical interrupt sources shared by the family (mutually exclusive).
	vl53SourceLevelLow       uint8 = 1
	vl53SourceLevelHigh      uint8 = 2
	vl53SourceOutOfWindow    uint8 = 3
	vl53SourceNewSampleReady uint8 = 4
	vl53SourceInWindow       uint8 = 5 // VL53L1X only
)

type vl53Base struct {
	conn       connection.Connection
	indexBytes int
	timeoutErr error

	// bus serializes multi-register sequences against the Full driver's
	// interrupt goroutine (a status read/clear must not interleave with them).
	bus sync.Mutex

	mu          sync.Mutex
	callback    func(uint8)
	unsubscribe func()
	pollPin     *connection.PollingInputPin
}

func newVL53Base(conn connection.Connection, indexBytes int, timeoutErr error) vl53Base {
	return vl53Base{conn: conn, indexBytes: indexBytes, timeoutErr: timeoutErr}
}

func (b *vl53Base) index(reg uint16) []byte {
	if b.indexBytes == 2 {
		return []byte{byte(reg >> 8), byte(reg)}
	}
	return []byte{byte(reg)}
}

func (b *vl53Base) writeBlock(reg uint16, data ...byte) error {
	return b.conn.Write(append(b.index(reg), data...))
}

func (b *vl53Base) readBlock(reg uint16, n int) ([]byte, error) {
	return b.conn.WriteRead(b.index(reg), n)
}

func (b *vl53Base) write8(reg uint16, value uint8) error {
	return b.writeBlock(reg, value)
}

func (b *vl53Base) read8(reg uint16) (uint8, error) {
	v, err := b.readBlock(reg, 1)
	if err != nil {
		return 0, err
	}
	return v[0], nil
}

func (b *vl53Base) write16(reg uint16, value uint16) error {
	return b.writeBlock(reg, byte(value>>8), byte(value))
}

func (b *vl53Base) read16(reg uint16) (uint16, error) {
	v, err := b.readBlock(reg, 2)
	if err != nil {
		return 0, err
	}
	return uint16(v[0])<<8 | uint16(v[1]), nil
}

func (b *vl53Base) write32(reg uint16, value uint32) error {
	return b.writeBlock(reg, byte(value>>24), byte(value>>16), byte(value>>8), byte(value))
}

func (b *vl53Base) read32(reg uint16) (uint32, error) {
	v, err := b.readBlock(reg, 4)
	if err != nil {
		return 0, err
	}
	return uint32(v[0])<<24 | uint32(v[1])<<16 | uint32(v[2])<<8 | uint32(v[3]), nil
}

// waitUntil polls predicate until it returns true; the chip's timeout error
// after 500 ms.
func (b *vl53Base) waitUntil(predicate func() (bool, error)) error {
	start := time.Now()
	for {
		ok, err := predicate()
		if err != nil {
			return err
		}
		if ok {
			return nil
		}
		if time.Since(start) > vl53Timeout {
			return b.timeoutErr
		}
	}
}

// bootWait drives XSHUT high (if the connection has an EnPin) and waits tBOOT.
func (b *vl53Base) bootWait() {
	if b.conn.EnPin() != nil {
		b.conn.Enable()
	}
	time.Sleep(vl53BootTime)
}

// setAddressReg writes a new 7-bit address to reg; invalidErr (no write)
// outside 0x08-0x77.
func (b *vl53Base) setAddressReg(reg uint16, address uint8, invalidErr error) error {
	if address < 0x08 || address > 0x77 {
		return invalidErr
	}
	return b.write8(reg, address&0x7F)
}

// subscribe delivers GPIO1 events (falling edge, active low) to callback.
// poll is the chip's locked read-and-clear; it returns the SOURCE value or 0.
// Without an IntPin a 5 ms polling pin is used.
func (b *vl53Base) subscribe(callback func(status uint8), poll func() (uint8, error)) error {
	b.mu.Lock()
	if b.unsubscribe != nil {
		b.unsubscribe()
		b.unsubscribe = nil
	}
	if b.pollPin != nil {
		_ = b.pollPin.Close()
		b.pollPin = nil
	}
	b.callback = callback
	b.mu.Unlock()

	pin := b.conn.IntPin()
	if pin == nil {
		p := connection.NewDefaultPollingInputPin()
		b.mu.Lock()
		b.pollPin = p
		b.mu.Unlock()
		pin = p
	}
	unsub := pin.OnEdge(connection.Falling, func() { b.handleEdge(poll) })
	b.mu.Lock()
	b.unsubscribe = unsub
	b.mu.Unlock()
	return nil
}

// unsubscribeAll detaches the edge handler, stops the polling pin and clears
// the callback.
func (b *vl53Base) unsubscribeAll() error {
	b.mu.Lock()
	unsub := b.unsubscribe
	b.unsubscribe = nil
	b.callback = nil
	p := b.pollPin
	b.pollPin = nil
	b.mu.Unlock()

	if unsub != nil {
		unsub()
	}
	if p != nil {
		return p.Close()
	}
	return nil
}

func (b *vl53Base) handleEdge(poll func() (uint8, error)) {
	status, err := poll()
	if err != nil || status == 0 {
		return
	}
	b.mu.Lock()
	cb := b.callback
	b.mu.Unlock()
	if cb != nil {
		cb(status)
	}
}
